### PR TITLE
Adding validation step for missing required ClusterTemplate's cluster…

### PR DESCRIPTION
**Pull request by edcdavid**
Adding validation step for missing required ClusterTemplate's clusterInstanceParameters fields:
- nodeClusterName
- oCloudSiteId
- policyTemplateParameters
- clusterInstanceParameters
Testing the for the presence of the required schema keys in the properties and in the required sections

### CODE REVIEW

File: internal/controllers/clustertemplate_controller.go [hunk 1]
```diff
@@ -327,6 +334,54 @@ func generateTemplateID(ctx context.Context, c client.Client, object *provisioni
...
 }
 
+// validateTemplateParameterSchema return true if the schema contained in the templateParameterSchema
+// field contains the required mandatory parameters
+// - nodeClusterName
+// - oCloudSiteId
+// - policyTemplateParameters
+// - clusterInstanceParameters
+func validateTemplateParameterSchema(object *provisioningv1alpha1.ClusterTemplate) error {
+	mandatoryParams := []string{"nodeClusterName",
```
Comment (irinamihai) [inline internal/controllers/clustertemplate_controller.go:344]: All these values will probably be standardized, so they can be made constants.

File: internal/controllers/clustertemplate_controller.go [hunk 2]
```diff
@@ -327,6 +334,54 @@ func generateTemplateID(ctx context.Context, c client.Client, object *provisioni
...
+		"oCloudSiteId",
+		"policyTemplateParameters",
+		"clusterInstanceParameters"}
+	if object.Spec.TemplateParameterSchema.Size() == 0 {
+		return utils.NewInputError("templateParameterSchema is present but empty:")
+	}
+	var missingParameter []string
+	for _, param := range mandatoryParams {
+		_, err := utils.ExtractSubSchema(object.Spec.TemplateParameterSchema.Raw, param)
+		if err != nil && strings.HasPrefix(err.Error(), fmt.Sprintf("subSchema %s does not exist:", param)) {
```
Comment (irinamihai) [inline internal/controllers/clustertemplate_controller.go:354]: Maybe refactor these lines to smth like:
```
if err != nil {
    if strings.HasPrefix(err.Error(), fmt.Sprintf("subSchema %s does not exist:", param)) {
        missingParameter = append(missingParameter, param)
    } else {
        return fmt.Errorf("error extracting subschema at key %s: %w", param, err)
    }
}
```

File: internal/controllers/clustertemplate_controller.go [hunk 3]
```diff
@@ -327,6 +334,54 @@ func generateTemplateID(ctx context.Context, c client.Client, object *provisioni
...
+	requiredList, err := utils.ExtractSchemaRequired(object.Spec.TemplateParameterSchema.Raw)
+	if err != nil {
+		return fmt.Errorf("error unmarshalling required subschema: %w", err)
+	}
+	for _, param := range mandatoryParams {
+		if !slices.Contains(requiredList, param) {
+			missingRequired = append(missingRequired, param)
+		}
+	}
+	validationFailureReason := fmt.Sprintf("failed to validate ClusterTemplate name:%s. ", object.Name)
```
Comment (irinamihai) [inline internal/controllers/clustertemplate_controller.go:370]: I think we can remove `name` from this

File: internal/controllers/clustertemplate_controller.go [hunk 4]
```diff
@@ -327,6 +334,54 @@ func generateTemplateID(ctx context.Context, c client.Client, object *provisioni
...
+	if len(missingParameter) != 0 {
+		validationFailureReason = fmt.Sprintf(" The following mandatory fields are missing: %s", strings.Join(missingParameter, ","))
+	}
+	if len(missingRequired) != 0 {
+		if len(missingParameter) != 0 {
```
Comment (irinamihai) [inline internal/controllers/clustertemplate_controller.go:375]: If you end the reason above with a `.` ()line 372 and start the one below with a `The`, you don't need this check.

File: internal/controllers/clustertemplate_controller.go (hunk 1 above)
Comment (Missxiaoguo) [reply internal/controllers/clustertemplate_controller.go:344]: There are existing constants for these values.

**Review by Missxiaoguo — COMMENTED** [review]
Could we also add the validation for the types? for example, 
`nodeClusterName` and `oCloudSiteId` should be of "type: string" and `policyTemplateParameters` and `clusterInstanceParameters` should be of "type: object". That's the key for validating the input types against the schema.

File: internal/controllers/clustertemplate_controller.go (hunk 1 above)
Comment (edcdavid) [reply internal/controllers/clustertemplate_controller.go]: Fixed

Comment (edcdavid) [reply internal/controllers/clustertemplate_controller.go]: Fixed

File: internal/controllers/clustertemplate_controller.go (hunk 2 above)
Comment (edcdavid) [reply internal/controllers/clustertemplate_controller.go]: Fixed

File: internal/controllers/clustertemplate_controller.go (hunk 3 above)
Comment (edcdavid) [reply internal/controllers/clustertemplate_controller.go]: Fixed

File: internal/controllers/clustertemplate_controller.go (hunk 4 above)
Comment (edcdavid) [reply internal/controllers/clustertemplate_controller.go]: Fixed.

**Comment by edcdavid** [comment]
> Could we also add the validation for the types? for example, `nodeClusterName` and `oCloudSiteId` should be of "type: string" and `policyTemplateParameters` and `clusterInstanceParameters` should be of "type: object". That's the key for validating the input types against the schema.

Fixed. Also added unit tests.

**Review by Missxiaoguo — COMMENTED** [review]
/lgtm

**Pull request merged**